phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixStatement.java

```diff
@@ -1993,19 +1992,31 @@ public void clearBatch() throws SQLException {
     @Override
     public int[] executeBatch() throws SQLException {
         int i = 0;
+        int[] returnCodes = new int [batch.size()];
+        Arrays.fill(returnCodes, -1);
+        boolean autoCommit = connection.getAutoCommit();
+        connection.setAutoCommit(false);
         try {
-            int[] returnCodes = new int [batch.size()];
             for (i = 0; i < returnCodes.length; i++) {
                 PhoenixPreparedStatement statement = batch.get(i);
-                returnCodes[i] = statement.execute(true) ? Statement.SUCCESS_NO_INFO : statement.getUpdateCount();
+                statement.executeForBatch();
+                returnCodes[i] = statement.getUpdateCount();
             }
             // Flush all changes in batch if auto flush is true
             flushIfNecessary();
             // If we make it all the way through, clear the batch
             clearBatch();
             return returnCodes;
-        } catch (Throwable t) {
-            throw new BatchUpdateExecution(t,i);
+        } catch (SQLException t) {
+            throw new BatchUpdateException(returnCodes, t);
```

Comment:
On exceptions shouldn't we set all the return codes to `Statement.EXECUTE_FAILED` since we don't know which individual query failed ?

Reply:
First of all, please note that I am actually initializing the array with `-1` which is currently an invalid value. I know that it is not formally reserved as an invalid value, but JDBC specifically leaves this out and uses `-2` for `SUCCESS_NO_INFO` and `-3` for `EXECUTE_FAILED` so it seems like they had this in their mind and so seems to be safe.

Regarding your thought, at this location, we would know which statement failed, so we could very well set that index alone to `EXECUTE_FAILED`, however, if the exception occurs in commit, you are right that we won't know. However, this is going to be the case even when `autocommit` is `false` so I think there is benefit in keeping the behavior same here as well.

Reply:
As mentioned in another thread, now `returnCodes` will exist only when for loop fails.